### DISCUSSION TITLE
Other fonts available?

### DISCUSSION THREAD

**Discussion by theissondergaard** [mrcodetastic/ESP32-HUB75-MatrixPanel-DMA#270] [Q&A] [answered]
I've started using the text-capabilities with the matrixpanel. But I can only make 1 font work, the FreeSansBold12pt7b which is used in the ChainedPanels example, using the virtualDisp. All attempts at other fonts fail to compile, saying the font wasn't declared in the scope. Any tips?

**Comment by ergindemir** [comment, marked as answer]
You have to add the header file for each font. Maybe that working font was added somewhere in the example.

**Reply by theissondergaard** [reply]
Aaaahhh, makes perfect sense. And now I see that include in the ESP32-VirtualMatrixPanel-I2S-DMA.h - I had been staring at the ESP32-HUB75-MatrixPanel-I2S-DMA.h :) Thanks for helping out a newbie!

**Reply by ergindemir** [reply]
Last time I tried custom fonts they were all at random locations unlike the default font. I am not sure if it is a library or GFX issue. Please let us know if you manage to locate them without any tweaks.

**Reply by GrimorJander** [reply]
I am facing the same problems to change font.
Following this thread, I have changed the "dma_display" classes to "virtualDisp->" just like in the Chained panel example. If I try to load a new font I get a compilation error except with the mentioned virtualDisp->setFont(&FreeSansBold12pt7b);

The question is should I include the Adafruit GX Font library at the beginning of the sketch? or only include the font.h as a library?
Shoul I modify ESP32-VirtualMatrixPanel-I2S-DMA.h in order to include each new font?
Where is default font located ? is FreeSansBold12pt7b default font?
I find the default font very bad for scrolling text

**Reply by GrimorJander** [reply]
Anyone?

**Reply by board707** [reply]
You should locate and include font header file for each font you using.
Not just a fonts.h, but something like 
#include "FreeSansBold12pt7b.h" 
for using the FreeSansBold12pt7b font

**Reply by GrimorJander** [reply]
![IMG_20221017_232609_edit_787208499226755](https://user-images.githubusercontent.com/27100419/196286953-69489e0b-c0c0-4ead-b7ca-cdc0c0ae38f4.jpg)
> You should locate and include font header file for each font you using. Not just a fonts.h, but something like #include "FreeSansBold12pt7b.h" for using the FreeSansBold12pt7b font

Thanks @board707 . Now i can test all font package. But i´m getting strange noise / color pixel blinking

`#include <Arduino.h>

#include <Adafruit_GFX.h>

#include "Fonts/FreeSerifBold9pt7b.h"

// Assume we have four 64x32 panels daizy-chained and ESP32 attached to the bottom right corner
#define PANEL_RES_X 64 // Number of pixels wide of each INDIVIDUAL panel module.
#define PANEL_RES_Y 32 // Number of pixels tall of each INDIVIDUAL panel module.

#define NUM_ROWS 1 // Number of rows of chained INDIVIDUAL PANELS
#define NUM_COLS 1 // Number of INDIVIDUAL PANELS per ROW
#define PANEL_CHAIN NUM_ROWS*NUM_COLS // total number of panels chained one to another

// Change this to your needs, for details on VirtualPanel pls see ChainedPanels example
#define SERPENT true
#define TOPDOWN true

#define VPANEL_W PANEL_RES_XNUM_COLS // Kosso: All Pattern files have had the MATRIX_WIDTH and MATRIX_HEIGHT replaced by these.
#define VPANEL_H PANEL_RES_YNUM_ROWS //

#include <ESP32-VirtualMatrixPanel-I2S-DMA.h>
String msg = "hello world";
// placeholder for the matrix object
MatrixPanel_I2S_DMA *matrix = nullptr;

// placeholder for the virtual display object
VirtualMatrixPanel *virtualDisp = nullptr;

void setup()
{
  // Setup serial interface
  Serial.begin(115200);
  delay(250);

  // Configure your matrix setup here
  HUB75_I2S_CFG mxconfig(PANEL_RES_X, PANEL_RES_Y, PANEL_CHAIN);

  // in case that we use panels based on FM6126A chip, we can change that
  mxconfig.driver = HUB75_I2S_CFG::FM6126A;

   // OK, now we can create our matrix object
  matrix = new MatrixPanel_I2S_DMA(mxconfig);

  // let's adjust default brightness to about 75%
  matrix->setBrightness8(96); // range is 0-255, 0 - 0%, 255 - 100%

  // Allocate memory and start DMA display
  if ( not matrix->begin() )
    Serial.println("****** !KABOOM! I2S memory allocation failed ***********");

  virtualDisp = new VirtualMatrixPanel((*matrix), NUM_ROWS, NUM_COLS, PANEL_RES_X, PANEL_RES_Y, SERPENT, TOPDOWN);
  virtualDisp->setTextWrap(false);
}

void loop() {
  {
    {
      virtualDisp->fillScreen(0);
    }
  }

  virtualDisp->setFont(&FreeSerifBold9pt7b);
  virtualDisp->setTextColor(virtualDisp->color333(7, 0, 4));
  virtualDisp->setCursor(0, 21);
  virtualDisp->setTextSize(1);
  virtualDisp->print(msg) ;
  delay(2550);
}`

**Comment by board707** [comment]
>  But i´m getting strange noise / color pixel blinking

When you used the default font, was there no noise?

**Reply by GrimorJander** [reply]
There is absolutely no noise with the default font. Also, when I scroll is even worse, the letters remain at the top of the display and an entire illuminated strip, I could not read any of the message

**Reply by GrimorJander** [reply]
Scrolling default font:
![IMG_20221018_192103](https://user-images.githubusercontent.com/27100419/196502275-4ada52a7-9312-4233-a127-e2fa527ab113.jpg)
GFX scrolling font:
![IMG_20221018_192450](https://user-images.githubusercontent.com/27100419/196502429-5a2300ee-1156-42c7-b0d1-0116e60f2a73.jpg)

**Reply by GrimorJander** [reply]
My fault. I forgot to fillscreen (myblack) between text X position. Now everything it's OK.
Listo! 
![IMG_20221018_233034](https://user-images.githubusercontent.com/27100419/196549146-d67eec62-6692-475f-aa85-b6dab9904d1f.jpg)

**Comment by board707** [comment]
My congratulations.

**Comment by primus192** [comment]
The one can use [this library](https://github.com/olikraus/U8g2_for_Adafruit_GFX) to benefit from u8g2 massive list of fonts and international diacritics / UTF-8 printing and line width calculation. I am using this library on top of virtualDisplay. This snippet may be of use:
```
#include <u8g2_for_Adafruit_GFX.h>

#define R1_PIN 0 //25
#define G1_PIN 18 //33
#define B1_PIN 2
#define R2_PIN 14
#define G2_PIN 12
#define B2_PIN 13
#define A_PIN 23
#define B_PIN 19
#define C_PIN 5
#define D_PIN 17
#define E_PIN 32
#define LAT_PIN 4
#define OE_PIN 15
#define CLK_PIN 16

#define SERPENT true
#define TOPDOWN false
#define NUM_ROWS 2
#define NUM_COLS 3
#define PANEL_RES_X 64      // Number of pixels wide of each INDIVIDUAL panel module. 
#define PANEL_RES_Y 64     // Number of pixels tall of each INDIVIDUAL panel module.
#define PANEL_CHAIN NUM_ROWS*NUM_COLS      // Total number of panels chained one to another

#include <ESP32-VirtualMatrixPanel-I2S-DMA.h>

MatrixPanel_I2S_DMA *dma_display = nullptr;
VirtualMatrixPanel  *virtualDisp = nullptr;
U8G2_FOR_ADAFRUIT_GFX *u8g2;

void setup() {
HUB75_I2S_CFG::i2s_pins _pins={R1_PIN, G1_PIN, B1_PIN, R2_PIN, G2_PIN, B2_PIN, A_PIN, B_PIN, C_PIN, D_PIN, E_PIN, LAT_PIN, OE_PIN, CLK_PIN};
    HUB75_I2S_CFG mxconfig(
                PANEL_RES_X,   // width
                PANEL_RES_Y,   // height
                PANEL_CHAIN,   // chain length
            _pins,   // pin mapping
    HUB75_I2S_CFG::FM6124      // driver chip
);
    mxconfig.i2sspeed = mxconfig.HZ_20M;

    dma_display = new MatrixPanel_I2S_DMA(mxconfig);
    dma_display->begin();
    virtualDisp = new VirtualMatrixPanel((*dma_display), NUM_ROWS, NUM_COLS, PANEL_RES_X, PANEL_RES_Y, SERPENT, TOPDOWN);
    u8g2 = new U8G2_FOR_ADAFRUIT_GFX();
    u8g2->begin(*virtualDisp);
}
```
Then we can use u8g2 functions to print etc.
